nvdaHelper/local/UIAEventLimiter/rateLimitedEventHandler.cpp

```diff
@@ -0,0 +1,227 @@
+/*
+This file is a part of the NVDA project.
+URL: http://github.com/nvaccess/nvda/
+Copyright 2023 NV Access Limited.
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License version 2.0, as published by
+    the Free Software Foundation.
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+This license can be found at:
+http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+*/
+
+#include <windows.h>
+#include <uiautomation.h>
+#include <numeric>
+#include <variant>
+#include <map>
+#include <vector>
+#include <functional>
+#include <mutex>
+#include <atomic>
+#include <atlcomcli.h>
+#include <comutil.h>
+#include <common/log.h>
+#include "eventRecord.h"
+#include "rateLimitedEventHandler.h"
+
+template<EventRecordConstraints EventRecordClass, typename... EventRecordArgTypes>
+HRESULT RateLimitedEventHandler::queueEvent(EventRecordArgTypes&&... args) {
+	LOG_DEBUG(L"RateLimitedUIAEventHandler::queueEvent called");
+	bool needsFlush = false;
+	const unsigned int flushTimeMS = (EventRecordClass::isCoalesceable)?30:0;
```

Comment:
The flush time is tricky. On one hand, we want rate limiting to be useful. On the other hand, this could impact perceived performance. In particular, property change events on the focus can be relevant to immediate user interaction; e.g. value changes on a focused volume slider. The response is likely to be slower than 30 ms because of timer resolution, additional time to queue the UIA event to NVDA core, etc.

I think the gains outweigh the losses here, but this is something we should at least consider for future optimisation. Perhaps we don't coalesce certain property changes on the focus or something like that?

Reply:
I think it would be worth then only doing coalescing for automation events, and not propertyChange events in this case. Not hard to change. Noting that the main motivation for this PR was textchange automation events. I think I should also add an extra variable to track if a flush message has been posted due to a non-coalesced event, and not post any more until the next flush. Currently if there are a whole bunch of non-coalesced events in a row, we'll keep flooding NVDA with pointless flush messages.

Reply:
Doing this only for automation events seems reasonable for this first PR if you prefer. However, I could see situations where property change event flooding could be a real issue. Task Manager is a good example. In that case, I think it does make sense to coalesce property change events except for the focus.

Yeah, the pointless flush messages thing is definitely worth fixing. I hadn't considered that.

Reply:
A couple of points:
1. PropertyChange events are only coalesced if their runtimeIDs match, and they are the same property (property ID). I.e. two name property changes on the same element are coalesced. I'm not too sure that this pr will really help task manager as I'm thinking its many properties across many list items.
2. Selective UIA registration is used in NVDA by default on Windows 11 now. thus there are hardly any property changes on elements other than the focus or its ancestors. I'd love to have this on by default in Windows 10 also, but apparently there are some bugs in Task list / emoji picker which stop us from doing that. @josephsl  @codeofdusk  ?
3. I've now added a commit that stops propertyChange events from being coalesced. and also a commit which stops the pointless flush messages.